src/main/java/appeng/me/service/helpers/NetworkCraftingProviders.java

```diff
@@ -0,0 +1,170 @@
+    private final Map<AEKey, Integer> emitableItems = new HashMap<>();
+
+    public void addProvider(IGridNode node) {
+        var provider = node.getService(ICraftingProvider.class);
+        if (provider != null) {
+            if (craftingProviders.containsKey(node)) {
+                throw new IllegalArgumentException("Duplicate crafting provider registration for node " + node);
+            }
+            var state = new ProviderState(provider);
+            state.mount(this);
+            craftingProviders.put(node, state);
+        }
+    }
+
+    public void removeProvider(IGridNode node) {
+        var provider = node.getService(ICraftingProvider.class);
+        if (provider != null) {
+            if (!craftingProviders.containsKey(node)) {
+                throw new IllegalArgumentException(
+                        "Attempted to remove nonexisting crafting provider for node " + node);
+            }
+            var state = craftingProviders.remove(node);
+            state.unmount(this);
+        }
+    }
+
+    public <T extends AEKey> Set<T> getCraftables(IStorageChannel<T> channel) {
+        var result = new HashSet<T>();
+
+        // add craftable items!
+        for (var stack : this.craftableItems.keySet()) {
+            if (stack.getChannel() == channel) {
+                result.add(stack.cast(channel));
+            }
+        }
+
+        for (var stack : this.emitableItems.keySet()) {
+            if (stack.getChannel() == channel) {
+                result.add(stack.cast(channel));
+            }
+        }
+
+        return result;
+    }
+
+    public Collection<IPatternDetails> getCraftingFor(AEKey whatToCraft) {
+        var patterns = this.craftableItems.get(whatToCraft);
+        if (patterns != null) {
+            return Collections.unmodifiableCollection(patterns.keySet());
+        }
+        return Collections.emptyList();
+    }
+
+    @Nullable
+    public AEKey getFuzzyCraftable(AEKey whatToCraft, Predicate<AEKey> filter) {
+        for (var fuzzy : craftableItemsList.findFuzzy(whatToCraft, FuzzyMode.IGNORE_ALL)) {
+            if (filter.test(fuzzy.getKey())) {
+                return fuzzy.getKey();
+            }
+        }
+        return null;
+    }
+
+    public boolean canEmitFor(AEKey someItem) {
+        return this.emitableItems.containsKey(someItem);
+    }
+
+    public Iterable<ICraftingProvider> getMediums(IPatternDetails key) {
+        var mediumList = this.craftingMethods.get(key);
+        if (mediumList != null) {
+            return mediumList.cycleIterable;
+        } else {
+            return Collections.emptyList();
+        }
+    }
+
+    private static class CraftingProviderList {
+        private final List<ICraftingProvider> providers = new ArrayList<>();
+        /**
+         * Cycling iterator for round-robin.
+         */
+        private final Iterator<ICraftingProvider> cycleIterator = Iterators.cycle(providers);
```

Comment:
Need to rebuild the iterator when the list changes due to an added pattern, otherwise we risk encountering a CME.